{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HW4 â€“ Statistics and Data Analysis\n",
    "### Differential Gene Expression in Acute Myocardial Infraction "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center> \n",
    "    Student IDs: 204818181, 329827190\n",
    "    </center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene expression describes the process in which genes that are coded in the DNA of living organisms are transcribed into mRNA. This is part of the bigger process in which genes are being copied (transcribed), processed, translated and modified into the final product, usually a protein. Gene expression profiling measures the levels at which mRNA molecules pertaining to the genes profiled are observed in a sample.\n",
    "In this exercise, we will perform guided analysis, comparing expression profiles of circulating endothelial cells (CECs) in patients with acute myocardial infraction to CECs in healthy controls. A comparison of two sample classes. You will then select one more gene expression dataset and perform your own analysis there.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data set was taken from:\n",
    "\n",
    "1)\tDataset record in NCBI:\n",
    "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE66360\n",
    "\n",
    "2)\tPublished paper: Muse et al, Sci Rep 2017 https://www.nature.com/articles/s41598-017-12166-0\n",
    "\n",
    "We extracted the data matrix and provide it as a separate csv attachment (link to download - https://drive.google.com/file/d/1-mu1J2wnlDIVNzoCXw20jJQ9r-h1jGrZ/view ). The csv file needs to be pre-processed before moving to the main analysis steps. Some information should be removed but make sure that you keep all information that is important for the analysis.\n",
    "The paper describes a study that seeks to develop an expression-based signature that can detect AMI in patients in a non-invasive manner, by profiling CECs. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Skipped the first 60 rows of data as they are invalid\n",
    "dataset = pd.read_csv(\"data.csv\", skiprows=58, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>H</th>\n",
       "      <th>...</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "      <th>M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ID_REF</td>\n",
       "      <td>GSM1620819</td>\n",
       "      <td>GSM1620820</td>\n",
       "      <td>GSM1620821</td>\n",
       "      <td>GSM1620822</td>\n",
       "      <td>GSM1620823</td>\n",
       "      <td>GSM1620824</td>\n",
       "      <td>GSM1620825</td>\n",
       "      <td>GSM1620826</td>\n",
       "      <td>GSM1620827</td>\n",
       "      <td>...</td>\n",
       "      <td>GSM1620908</td>\n",
       "      <td>GSM1620909</td>\n",
       "      <td>GSM1620910</td>\n",
       "      <td>GSM1620911</td>\n",
       "      <td>GSM1620912</td>\n",
       "      <td>GSM1620913</td>\n",
       "      <td>GSM1620914</td>\n",
       "      <td>GSM1620915</td>\n",
       "      <td>GSM1620916</td>\n",
       "      <td>GSM1620917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>5.866377893</td>\n",
       "      <td>4.948736331</td>\n",
       "      <td>5.148384393</td>\n",
       "      <td>5.064418945</td>\n",
       "      <td>5.017832878</td>\n",
       "      <td>5.116153518</td>\n",
       "      <td>5.431328058</td>\n",
       "      <td>5.235270857</td>\n",
       "      <td>5.773528455</td>\n",
       "      <td>...</td>\n",
       "      <td>5.419481538</td>\n",
       "      <td>5.057716465</td>\n",
       "      <td>5.996493392</td>\n",
       "      <td>5.343132759</td>\n",
       "      <td>5.558892254</td>\n",
       "      <td>6.472517225</td>\n",
       "      <td>5.678815851</td>\n",
       "      <td>5.653286378</td>\n",
       "      <td>6.013841046</td>\n",
       "      <td>5.465333944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1053_at</td>\n",
       "      <td>8.22579012</td>\n",
       "      <td>7.525865369</td>\n",
       "      <td>7.764636529</td>\n",
       "      <td>7.198461482</td>\n",
       "      <td>7.831450258</td>\n",
       "      <td>7.203591859</td>\n",
       "      <td>7.694550756</td>\n",
       "      <td>7.760259212</td>\n",
       "      <td>8.279814404</td>\n",
       "      <td>...</td>\n",
       "      <td>7.226347747</td>\n",
       "      <td>7.105537863</td>\n",
       "      <td>5.354105386</td>\n",
       "      <td>8.271499725</td>\n",
       "      <td>5.96334574</td>\n",
       "      <td>8.261421952</td>\n",
       "      <td>6.925752665</td>\n",
       "      <td>7.918424183</td>\n",
       "      <td>7.442701377</td>\n",
       "      <td>9.225221352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>117_at</td>\n",
       "      <td>6.17973231</td>\n",
       "      <td>6.628137025</td>\n",
       "      <td>5.859151477</td>\n",
       "      <td>5.974135101</td>\n",
       "      <td>6.793079158</td>\n",
       "      <td>6.43522914</td>\n",
       "      <td>6.320546126</td>\n",
       "      <td>6.48359047</td>\n",
       "      <td>6.387779205</td>\n",
       "      <td>...</td>\n",
       "      <td>6.789500767</td>\n",
       "      <td>7.441939912</td>\n",
       "      <td>7.026928573</td>\n",
       "      <td>6.003972814</td>\n",
       "      <td>7.474325713</td>\n",
       "      <td>6.367759272</td>\n",
       "      <td>6.835755831</td>\n",
       "      <td>7.577034915</td>\n",
       "      <td>7.147417202</td>\n",
       "      <td>6.48112813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>121_at</td>\n",
       "      <td>6.179477974</td>\n",
       "      <td>6.58288903</td>\n",
       "      <td>6.602134766</td>\n",
       "      <td>6.545904723</td>\n",
       "      <td>5.911542321</td>\n",
       "      <td>6.28542026</td>\n",
       "      <td>6.562315839</td>\n",
       "      <td>6.345887555</td>\n",
       "      <td>6.27280582</td>\n",
       "      <td>...</td>\n",
       "      <td>6.513436777</td>\n",
       "      <td>6.317522639</td>\n",
       "      <td>7.413453376</td>\n",
       "      <td>6.809103167</td>\n",
       "      <td>5.949091368</td>\n",
       "      <td>6.440978114</td>\n",
       "      <td>6.955883278</td>\n",
       "      <td>6.499215058</td>\n",
       "      <td>6.587065112</td>\n",
       "      <td>6.897590966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54672</td>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>3.139333661</td>\n",
       "      <td>3.039961017</td>\n",
       "      <td>3.220109351</td>\n",
       "      <td>3.263781541</td>\n",
       "      <td>3.138524252</td>\n",
       "      <td>3.043078536</td>\n",
       "      <td>3.231107638</td>\n",
       "      <td>3.304715967</td>\n",
       "      <td>3.384011442</td>\n",
       "      <td>...</td>\n",
       "      <td>3.525011316</td>\n",
       "      <td>3.309030335</td>\n",
       "      <td>3.663399948</td>\n",
       "      <td>3.32400503</td>\n",
       "      <td>4.619261018</td>\n",
       "      <td>5.558780351</td>\n",
       "      <td>3.416999272</td>\n",
       "      <td>3.699181146</td>\n",
       "      <td>3.48527965</td>\n",
       "      <td>4.211921779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54673</td>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>2.7470148</td>\n",
       "      <td>2.63860588</td>\n",
       "      <td>2.495261011</td>\n",
       "      <td>2.544198973</td>\n",
       "      <td>2.506629527</td>\n",
       "      <td>2.817981885</td>\n",
       "      <td>2.525060286</td>\n",
       "      <td>2.91727791</td>\n",
       "      <td>2.643849254</td>\n",
       "      <td>...</td>\n",
       "      <td>2.930794418</td>\n",
       "      <td>2.745955351</td>\n",
       "      <td>3.165167607</td>\n",
       "      <td>2.899104156</td>\n",
       "      <td>4.066775612</td>\n",
       "      <td>5.429436708</td>\n",
       "      <td>2.834741433</td>\n",
       "      <td>3.24968301</td>\n",
       "      <td>2.71141329</td>\n",
       "      <td>3.346834702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54674</td>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>2.651554479</td>\n",
       "      <td>2.643615067</td>\n",
       "      <td>2.626028059</td>\n",
       "      <td>2.504957719</td>\n",
       "      <td>2.609707404</td>\n",
       "      <td>2.521610278</td>\n",
       "      <td>2.840601479</td>\n",
       "      <td>2.509774164</td>\n",
       "      <td>2.696820467</td>\n",
       "      <td>...</td>\n",
       "      <td>2.788852359</td>\n",
       "      <td>2.725246769</td>\n",
       "      <td>3.143400613</td>\n",
       "      <td>2.76875475</td>\n",
       "      <td>3.546116866</td>\n",
       "      <td>4.487940292</td>\n",
       "      <td>2.782355764</td>\n",
       "      <td>3.056726837</td>\n",
       "      <td>2.833818355</td>\n",
       "      <td>3.06954169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54675</td>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>3.413429017</td>\n",
       "      <td>3.399062751</td>\n",
       "      <td>3.539969696</td>\n",
       "      <td>3.396225335</td>\n",
       "      <td>3.43472012</td>\n",
       "      <td>3.516089622</td>\n",
       "      <td>3.514949337</td>\n",
       "      <td>3.475952299</td>\n",
       "      <td>3.495198562</td>\n",
       "      <td>...</td>\n",
       "      <td>3.701271488</td>\n",
       "      <td>3.456716807</td>\n",
       "      <td>4.127459216</td>\n",
       "      <td>3.668326993</td>\n",
       "      <td>4.889324389</td>\n",
       "      <td>4.254285925</td>\n",
       "      <td>3.882079933</td>\n",
       "      <td>3.649826789</td>\n",
       "      <td>3.523410023</td>\n",
       "      <td>4.108475085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>54676</td>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>3.140611771</td>\n",
       "      <td>3.132691213</td>\n",
       "      <td>3.698396097</td>\n",
       "      <td>3.638611866</td>\n",
       "      <td>3.577649067</td>\n",
       "      <td>3.774955506</td>\n",
       "      <td>3.526497456</td>\n",
       "      <td>3.626839179</td>\n",
       "      <td>3.599541762</td>\n",
       "      <td>...</td>\n",
       "      <td>4.067019343</td>\n",
       "      <td>3.638374125</td>\n",
       "      <td>3.721337802</td>\n",
       "      <td>3.517545476</td>\n",
       "      <td>4.258245885</td>\n",
       "      <td>5.375810973</td>\n",
       "      <td>3.744138165</td>\n",
       "      <td>4.021536308</td>\n",
       "      <td>3.81256443</td>\n",
       "      <td>4.59212217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54676 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                Class            H            H            H            H  \\\n",
       "1               ID_REF   GSM1620819   GSM1620820   GSM1620821   GSM1620822   \n",
       "2            1007_s_at  5.866377893  4.948736331  5.148384393  5.064418945   \n",
       "3              1053_at   8.22579012  7.525865369  7.764636529  7.198461482   \n",
       "4               117_at   6.17973231  6.628137025  5.859151477  5.974135101   \n",
       "5               121_at  6.179477974   6.58288903  6.602134766  6.545904723   \n",
       "...                ...          ...          ...          ...          ...   \n",
       "54672   AFFX-ThrX-5_at  3.139333661  3.039961017  3.220109351  3.263781541   \n",
       "54673   AFFX-ThrX-M_at    2.7470148   2.63860588  2.495261011  2.544198973   \n",
       "54674  AFFX-TrpnX-3_at  2.651554479  2.643615067  2.626028059  2.504957719   \n",
       "54675  AFFX-TrpnX-5_at  3.413429017  3.399062751  3.539969696  3.396225335   \n",
       "54676  AFFX-TrpnX-M_at  3.140611771  3.132691213  3.698396097  3.638611866   \n",
       "\n",
       "0                H            H            H            H            H  ...  \\\n",
       "1       GSM1620823   GSM1620824   GSM1620825   GSM1620826   GSM1620827  ...   \n",
       "2      5.017832878  5.116153518  5.431328058  5.235270857  5.773528455  ...   \n",
       "3      7.831450258  7.203591859  7.694550756  7.760259212  8.279814404  ...   \n",
       "4      6.793079158   6.43522914  6.320546126   6.48359047  6.387779205  ...   \n",
       "5      5.911542321   6.28542026  6.562315839  6.345887555   6.27280582  ...   \n",
       "...            ...          ...          ...          ...          ...  ...   \n",
       "54672  3.138524252  3.043078536  3.231107638  3.304715967  3.384011442  ...   \n",
       "54673  2.506629527  2.817981885  2.525060286   2.91727791  2.643849254  ...   \n",
       "54674  2.609707404  2.521610278  2.840601479  2.509774164  2.696820467  ...   \n",
       "54675   3.43472012  3.516089622  3.514949337  3.475952299  3.495198562  ...   \n",
       "54676  3.577649067  3.774955506  3.526497456  3.626839179  3.599541762  ...   \n",
       "\n",
       "0                M            M            M            M            M  \\\n",
       "1       GSM1620908   GSM1620909   GSM1620910   GSM1620911   GSM1620912   \n",
       "2      5.419481538  5.057716465  5.996493392  5.343132759  5.558892254   \n",
       "3      7.226347747  7.105537863  5.354105386  8.271499725   5.96334574   \n",
       "4      6.789500767  7.441939912  7.026928573  6.003972814  7.474325713   \n",
       "5      6.513436777  6.317522639  7.413453376  6.809103167  5.949091368   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "54672  3.525011316  3.309030335  3.663399948   3.32400503  4.619261018   \n",
       "54673  2.930794418  2.745955351  3.165167607  2.899104156  4.066775612   \n",
       "54674  2.788852359  2.725246769  3.143400613   2.76875475  3.546116866   \n",
       "54675  3.701271488  3.456716807  4.127459216  3.668326993  4.889324389   \n",
       "54676  4.067019343  3.638374125  3.721337802  3.517545476  4.258245885   \n",
       "\n",
       "0                M            M            M            M            M  \n",
       "1       GSM1620913   GSM1620914   GSM1620915   GSM1620916   GSM1620917  \n",
       "2      6.472517225  5.678815851  5.653286378  6.013841046  5.465333944  \n",
       "3      8.261421952  6.925752665  7.918424183  7.442701377  9.225221352  \n",
       "4      6.367759272  6.835755831  7.577034915  7.147417202   6.48112813  \n",
       "5      6.440978114  6.955883278  6.499215058  6.587065112  6.897590966  \n",
       "...            ...          ...          ...          ...          ...  \n",
       "54672  5.558780351  3.416999272  3.699181146   3.48527965  4.211921779  \n",
       "54673  5.429436708  2.834741433   3.24968301   2.71141329  3.346834702  \n",
       "54674  4.487940292  2.782355764  3.056726837  2.833818355   3.06954169  \n",
       "54675  4.254285925  3.882079933  3.649826789  3.523410023  4.108475085  \n",
       "54676  5.375810973  3.744138165  4.021536308   3.81256443   4.59212217  \n",
       "\n",
       "[54676 rows x 100 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Setting up data DataFrame\n",
    "header = dataset.iloc[0]\n",
    "data = dataset[1:] \n",
    "data.columns = header\n",
    "type(data)\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A.\tHigh level description of the data and some pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1)\tHow many genes profiled? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 54675 profiled genes\n"
     ]
    }
   ],
   "source": [
    "print('There are {} profiled genes'.format(data.shape[0] - 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2)\tHow many samples (subjects/patients) in total?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 99 patients\n"
     ]
    }
   ],
   "source": [
    "print('There are {} patients'.format(data.shape[1] - 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3)\tHow many samples in each class?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "      <th>ID_REF</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>H</td>\n",
       "      <td>GSM1620819</td>\n",
       "      <td>5.866377893</td>\n",
       "      <td>8.22579012</td>\n",
       "      <td>6.17973231</td>\n",
       "      <td>6.179477974</td>\n",
       "      <td>2.792047952</td>\n",
       "      <td>9.290410779</td>\n",
       "      <td>6.998285145</td>\n",
       "      <td>5.348943925</td>\n",
       "      <td>...</td>\n",
       "      <td>13.25732501</td>\n",
       "      <td>13.26623454</td>\n",
       "      <td>14.57122985</td>\n",
       "      <td>14.27656808</td>\n",
       "      <td>4.504659469</td>\n",
       "      <td>3.139333661</td>\n",
       "      <td>2.7470148</td>\n",
       "      <td>2.651554479</td>\n",
       "      <td>3.413429017</td>\n",
       "      <td>3.140611771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>H</td>\n",
       "      <td>GSM1620820</td>\n",
       "      <td>4.948736331</td>\n",
       "      <td>7.525865369</td>\n",
       "      <td>6.628137025</td>\n",
       "      <td>6.58288903</td>\n",
       "      <td>2.69571445</td>\n",
       "      <td>9.462275035</td>\n",
       "      <td>6.558214949</td>\n",
       "      <td>5.410884095</td>\n",
       "      <td>...</td>\n",
       "      <td>13.59386023</td>\n",
       "      <td>13.51701336</td>\n",
       "      <td>14.73883363</td>\n",
       "      <td>14.34123318</td>\n",
       "      <td>4.409225776</td>\n",
       "      <td>3.039961017</td>\n",
       "      <td>2.63860588</td>\n",
       "      <td>2.643615067</td>\n",
       "      <td>3.399062751</td>\n",
       "      <td>3.132691213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>H</td>\n",
       "      <td>GSM1620821</td>\n",
       "      <td>5.148384393</td>\n",
       "      <td>7.764636529</td>\n",
       "      <td>5.859151477</td>\n",
       "      <td>6.602134766</td>\n",
       "      <td>2.580181122</td>\n",
       "      <td>9.116776316</td>\n",
       "      <td>6.851622539</td>\n",
       "      <td>5.254073031</td>\n",
       "      <td>...</td>\n",
       "      <td>13.37275868</td>\n",
       "      <td>13.41658291</td>\n",
       "      <td>14.62718054</td>\n",
       "      <td>14.31856805</td>\n",
       "      <td>4.310075612</td>\n",
       "      <td>3.220109351</td>\n",
       "      <td>2.495261011</td>\n",
       "      <td>2.626028059</td>\n",
       "      <td>3.539969696</td>\n",
       "      <td>3.698396097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>H</td>\n",
       "      <td>GSM1620822</td>\n",
       "      <td>5.064418945</td>\n",
       "      <td>7.198461482</td>\n",
       "      <td>5.974135101</td>\n",
       "      <td>6.545904723</td>\n",
       "      <td>2.712123845</td>\n",
       "      <td>9.134584186</td>\n",
       "      <td>7.357827486</td>\n",
       "      <td>5.242139865</td>\n",
       "      <td>...</td>\n",
       "      <td>13.49022556</td>\n",
       "      <td>13.49482728</td>\n",
       "      <td>14.75412393</td>\n",
       "      <td>14.3632965</td>\n",
       "      <td>4.439310699</td>\n",
       "      <td>3.263781541</td>\n",
       "      <td>2.544198973</td>\n",
       "      <td>2.504957719</td>\n",
       "      <td>3.396225335</td>\n",
       "      <td>3.638611866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>H</td>\n",
       "      <td>GSM1620823</td>\n",
       "      <td>5.017832878</td>\n",
       "      <td>7.831450258</td>\n",
       "      <td>6.793079158</td>\n",
       "      <td>5.911542321</td>\n",
       "      <td>2.723686912</td>\n",
       "      <td>9.582343511</td>\n",
       "      <td>7.283872601</td>\n",
       "      <td>5.219483996</td>\n",
       "      <td>...</td>\n",
       "      <td>13.23678545</td>\n",
       "      <td>13.33019244</td>\n",
       "      <td>14.54838931</td>\n",
       "      <td>14.26367675</td>\n",
       "      <td>4.562883241</td>\n",
       "      <td>3.138524252</td>\n",
       "      <td>2.506629527</td>\n",
       "      <td>2.609707404</td>\n",
       "      <td>3.43472012</td>\n",
       "      <td>3.577649067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>95</td>\n",
       "      <td>M</td>\n",
       "      <td>GSM1620913</td>\n",
       "      <td>6.472517225</td>\n",
       "      <td>8.261421952</td>\n",
       "      <td>6.367759272</td>\n",
       "      <td>6.440978114</td>\n",
       "      <td>3.73013912</td>\n",
       "      <td>7.355729877</td>\n",
       "      <td>6.206879166</td>\n",
       "      <td>6.217929861</td>\n",
       "      <td>...</td>\n",
       "      <td>6.601745953</td>\n",
       "      <td>5.94230757</td>\n",
       "      <td>7.290513508</td>\n",
       "      <td>8.988510881</td>\n",
       "      <td>5.567947288</td>\n",
       "      <td>5.558780351</td>\n",
       "      <td>5.429436708</td>\n",
       "      <td>4.487940292</td>\n",
       "      <td>4.254285925</td>\n",
       "      <td>5.375810973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>96</td>\n",
       "      <td>M</td>\n",
       "      <td>GSM1620914</td>\n",
       "      <td>5.678815851</td>\n",
       "      <td>6.925752665</td>\n",
       "      <td>6.835755831</td>\n",
       "      <td>6.955883278</td>\n",
       "      <td>2.914867562</td>\n",
       "      <td>7.944294937</td>\n",
       "      <td>6.638364166</td>\n",
       "      <td>5.992343792</td>\n",
       "      <td>...</td>\n",
       "      <td>14.2710373</td>\n",
       "      <td>14.16370633</td>\n",
       "      <td>14.88450605</td>\n",
       "      <td>14.79250553</td>\n",
       "      <td>4.902195229</td>\n",
       "      <td>3.416999272</td>\n",
       "      <td>2.834741433</td>\n",
       "      <td>2.782355764</td>\n",
       "      <td>3.882079933</td>\n",
       "      <td>3.744138165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>M</td>\n",
       "      <td>GSM1620915</td>\n",
       "      <td>5.653286378</td>\n",
       "      <td>7.918424183</td>\n",
       "      <td>7.577034915</td>\n",
       "      <td>6.499215058</td>\n",
       "      <td>3.013901455</td>\n",
       "      <td>8.563122679</td>\n",
       "      <td>7.379567509</td>\n",
       "      <td>5.766028609</td>\n",
       "      <td>...</td>\n",
       "      <td>13.89861618</td>\n",
       "      <td>13.68242475</td>\n",
       "      <td>14.88450605</td>\n",
       "      <td>14.78380144</td>\n",
       "      <td>5.239906199</td>\n",
       "      <td>3.699181146</td>\n",
       "      <td>3.24968301</td>\n",
       "      <td>3.056726837</td>\n",
       "      <td>3.649826789</td>\n",
       "      <td>4.021536308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>M</td>\n",
       "      <td>GSM1620916</td>\n",
       "      <td>6.013841046</td>\n",
       "      <td>7.442701377</td>\n",
       "      <td>7.147417202</td>\n",
       "      <td>6.587065112</td>\n",
       "      <td>2.729319786</td>\n",
       "      <td>8.699221635</td>\n",
       "      <td>6.598349589</td>\n",
       "      <td>5.577456268</td>\n",
       "      <td>...</td>\n",
       "      <td>13.84700646</td>\n",
       "      <td>13.55159338</td>\n",
       "      <td>14.85806628</td>\n",
       "      <td>14.7781659</td>\n",
       "      <td>4.556134169</td>\n",
       "      <td>3.48527965</td>\n",
       "      <td>2.71141329</td>\n",
       "      <td>2.833818355</td>\n",
       "      <td>3.523410023</td>\n",
       "      <td>3.81256443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>M</td>\n",
       "      <td>GSM1620917</td>\n",
       "      <td>5.465333944</td>\n",
       "      <td>9.225221352</td>\n",
       "      <td>6.48112813</td>\n",
       "      <td>6.897590966</td>\n",
       "      <td>3.151449425</td>\n",
       "      <td>8.194374083</td>\n",
       "      <td>6.133298023</td>\n",
       "      <td>6.193797335</td>\n",
       "      <td>...</td>\n",
       "      <td>14.08796502</td>\n",
       "      <td>13.86742522</td>\n",
       "      <td>14.88272796</td>\n",
       "      <td>14.79250553</td>\n",
       "      <td>5.573424723</td>\n",
       "      <td>4.211921779</td>\n",
       "      <td>3.346834702</td>\n",
       "      <td>3.06954169</td>\n",
       "      <td>4.108475085</td>\n",
       "      <td>4.59212217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows Ã— 54677 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0  Class      ID_REF    1007_s_at      1053_at       117_at       121_at  \\\n",
       "1      H  GSM1620819  5.866377893   8.22579012   6.17973231  6.179477974   \n",
       "2      H  GSM1620820  4.948736331  7.525865369  6.628137025   6.58288903   \n",
       "3      H  GSM1620821  5.148384393  7.764636529  5.859151477  6.602134766   \n",
       "4      H  GSM1620822  5.064418945  7.198461482  5.974135101  6.545904723   \n",
       "5      H  GSM1620823  5.017832878  7.831450258  6.793079158  5.911542321   \n",
       "..   ...         ...          ...          ...          ...          ...   \n",
       "95     M  GSM1620913  6.472517225  8.261421952  6.367759272  6.440978114   \n",
       "96     M  GSM1620914  5.678815851  6.925752665  6.835755831  6.955883278   \n",
       "97     M  GSM1620915  5.653286378  7.918424183  7.577034915  6.499215058   \n",
       "98     M  GSM1620916  6.013841046  7.442701377  7.147417202  6.587065112   \n",
       "99     M  GSM1620917  5.465333944  9.225221352   6.48112813  6.897590966   \n",
       "\n",
       "0     1255_g_at      1294_at      1316_at      1320_at  ...  \\\n",
       "1   2.792047952  9.290410779  6.998285145  5.348943925  ...   \n",
       "2    2.69571445  9.462275035  6.558214949  5.410884095  ...   \n",
       "3   2.580181122  9.116776316  6.851622539  5.254073031  ...   \n",
       "4   2.712123845  9.134584186  7.357827486  5.242139865  ...   \n",
       "5   2.723686912  9.582343511  7.283872601  5.219483996  ...   \n",
       "..          ...          ...          ...          ...  ...   \n",
       "95   3.73013912  7.355729877  6.206879166  6.217929861  ...   \n",
       "96  2.914867562  7.944294937  6.638364166  5.992343792  ...   \n",
       "97  3.013901455  8.563122679  7.379567509  5.766028609  ...   \n",
       "98  2.729319786  8.699221635  6.598349589  5.577456268  ...   \n",
       "99  3.151449425  8.194374083  6.133298023  6.193797335  ...   \n",
       "\n",
       "0  AFFX-r2-Ec-bioD-3_at AFFX-r2-Ec-bioD-5_at AFFX-r2-P1-cre-3_at  \\\n",
       "1           13.25732501          13.26623454         14.57122985   \n",
       "2           13.59386023          13.51701336         14.73883363   \n",
       "3           13.37275868          13.41658291         14.62718054   \n",
       "4           13.49022556          13.49482728         14.75412393   \n",
       "5           13.23678545          13.33019244         14.54838931   \n",
       "..                  ...                  ...                 ...   \n",
       "95          6.601745953           5.94230757         7.290513508   \n",
       "96           14.2710373          14.16370633         14.88450605   \n",
       "97          13.89861618          13.68242475         14.88450605   \n",
       "98          13.84700646          13.55159338         14.85806628   \n",
       "99          14.08796502          13.86742522         14.88272796   \n",
       "\n",
       "0  AFFX-r2-P1-cre-5_at AFFX-ThrX-3_at AFFX-ThrX-5_at AFFX-ThrX-M_at  \\\n",
       "1          14.27656808    4.504659469    3.139333661      2.7470148   \n",
       "2          14.34123318    4.409225776    3.039961017     2.63860588   \n",
       "3          14.31856805    4.310075612    3.220109351    2.495261011   \n",
       "4           14.3632965    4.439310699    3.263781541    2.544198973   \n",
       "5          14.26367675    4.562883241    3.138524252    2.506629527   \n",
       "..                 ...            ...            ...            ...   \n",
       "95         8.988510881    5.567947288    5.558780351    5.429436708   \n",
       "96         14.79250553    4.902195229    3.416999272    2.834741433   \n",
       "97         14.78380144    5.239906199    3.699181146     3.24968301   \n",
       "98          14.7781659    4.556134169     3.48527965     2.71141329   \n",
       "99         14.79250553    5.573424723    4.211921779    3.346834702   \n",
       "\n",
       "0  AFFX-TrpnX-3_at AFFX-TrpnX-5_at AFFX-TrpnX-M_at  \n",
       "1      2.651554479     3.413429017     3.140611771  \n",
       "2      2.643615067     3.399062751     3.132691213  \n",
       "3      2.626028059     3.539969696     3.698396097  \n",
       "4      2.504957719     3.396225335     3.638611866  \n",
       "5      2.609707404      3.43472012     3.577649067  \n",
       "..             ...             ...             ...  \n",
       "95     4.487940292     4.254285925     5.375810973  \n",
       "96     2.782355764     3.882079933     3.744138165  \n",
       "97     3.056726837     3.649826789     4.021536308  \n",
       "98     2.833818355     3.523410023      3.81256443  \n",
       "99      3.06954169     4.108475085      4.59212217  \n",
       "\n",
       "[99 rows x 54677 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We will transpose this data so it makes more sense to work with\n",
    "transpose_temp = data.transpose()\n",
    "transpose_temp = transpose_temp.reset_index()\n",
    "transpose_header = transpose_temp.iloc[0]\n",
    "transpose_data = transpose_temp[1:] \n",
    "transpose_data.columns = transpose_header\n",
    "transpose_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There 50 Hs and 49 Ms\n"
     ]
    }
   ],
   "source": [
    "print('There {0} Hs and {1} Ms'.format(list(transpose_data[\"Class\"].value_counts())[0],list(transpose_data[\"Class\"].value_counts())[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4)\tIf there are missing values, then remove the entire row (gene) from the data matrix. How many rows left now?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 54676 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "transpose_data.dropna(how='any',axis=0,inplace=True)\n",
    "print('There are {} genes'.format(transpose_data.shape[1] - 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B.\tWRS for differential expression (DE)\n",
    "\n",
    "1)\tConsider some gene, g. Under the null model (which assumes that for g there is no M vs H DE), what is the expected sum of ranks of gâ€™s expression levels measured for samples labeled M?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$E[RS(g)]= \\frac{B(N+1)}{2} = \\frac{B(g(M)+1)}{2} = \\frac{49Â·100}{2} =2450$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2)\tDenote this sum of ranks by RS(g). What is the minimal value, m, that RS(g) can take?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will take the sum of the first 49 ranks\n",
    "$$ \\sum^{49}_{i=1} RS(g) = \\frac{n(n+1)}{2} =\\frac{49(50)}{2} = 1225 $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3)\tUnder the null model, what is the probability of RS(g) = m? \n",
    "(provide a formula for this and explain it)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to find the probability of RS(g) = m, we will consider the number of class premutation from 99 patients. \n",
    "\n",
    "That is $ {99 \\choose 49} $ possible premutations. Then\n",
    "$$ P[RS(g) = m] = \\frac{1}{{99 \\choose 49}} $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4)\tUnder the null model, what is the probability of RS(g) = m+1? what is the probability of RS(g) = m+2? \n",
    "(provide formulas and explain them)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Under the null model, for $RS(g) = m + 1 $, we have\n",
    "$$ P[RS(g) = m] = \\frac{1}{{99 \\choose 49}} $$\n",
    "and from $RS(g) = m + 2 $, we have\n",
    "$$ P[RS(g) = m] = \\frac{2}{{99 \\choose 49}} $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5)\tDraw a histogram of the values of RS(g) in the dataset. Here g ranges over all genes in the data (after the clean-up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'rank'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-cd3d72646d47>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtranspose_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Class'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"M\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtranspose_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0mrank_sum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# ranked_df = transpose_data.iloc[:,2:].apply(rank_sum, axis=0)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-11-cd3d72646d47>\u001b[0m in \u001b[0;36mrank_sum\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mrank_sum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mrank\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrank\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mranks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtranspose_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Class'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"M\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtranspose_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Class'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"M\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'rank'"
     ]
    }
   ],
   "source": [
    "def rank_sum(gene):\n",
    "    rank = gene.rank()\n",
    "    print( ranks[transpose_data['Class'] == \"M\"].sum())\n",
    "\n",
    "transpose_data['Class'] == \"M\"\n",
    "for gene in transpose_data:\n",
    "    rank_sum(gene)\n",
    "    \n",
    "# ranked_df = transpose_data.iloc[:,2:].apply(rank_sum, axis=0)\n",
    "# ranked_df.hist(bins=30, density=True)\n",
    "# plt.title(\"RS(g) over all genes\")\n",
    "# plt.ylabel(\"Density\")\n",
    "# plt.xlabel(\"RS(g)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C. Differential Expression\n",
    "The purpose is to determine the statistical significance of differential expression (DE) observed for each gene in H vs M.\n",
    "Evaluate the DE in both one-sided directions for every gene, using both Student t-test and WRS test. \n",
    "Report the number of genes overexpressed in M vs H at a p-value better (â‰¤) than 0.05 and separately genes underexpressed in M vs H at a p-value of 0.05. For both directions use both a Student t-test and a WRS test.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D.\tCorrelations\n",
    "Select the 60 most significant genes from each one of the one-sided WRS DE lists you computed in 3c. Generate a set of 120 genes, D, which is the union of the above two sets.\n",
    "Compute Spearman rho correlations in all pairs within D (120 choose 2 numbers). \n",
    "\n",
    "1)\tWhat can you report about co-expression of genes in D (co-expression is inferred from the correlation of the expression levels of genes, across a set of samples)? Do we observe any significant co-expression? If so how many pairs, etc.\n",
    "\n",
    "2)\tWhat would have been advantages and disadvantages of computing co-expression for all genes in the study rather than only for genes in D? \n",
    "\n",
    "\n",
    "3)\tPerform the above steps on the same set D, but restrict attention only to samples labeled M. What do you see now? Can you explain this?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### E.\tPlots and Conclusions of the DE and correlation analysis\n",
    "\n",
    "1)\tConstruct the DE overabundance plots (blue and green lines as shown in class) for M vs H overexpression (higher expression levels in M) using WRS and t-test using the results you had computed in Section 3c. \n",
    "State, for each comparison, the number of genes, k, at which we observe:\n",
    "\n",
    "     a)\tFDR = 0.1\n",
    "     b)\tFDR = 0.05\n",
    "     c)\tFDR = 0.001\n",
    "\n",
    "If these events are not observed at any k, then make that statement.\n",
    "\n",
    "2)\tWhat can you say about the difference in results obtained in WRS vs those obtained by Student t-test?\n",
    "\n",
    "3)\tSelect any 3 differentially expressed genes, from D (which was defined in 3d), and produce a graphical representation of their expression patterns that demonstrates the observed DE.\n",
    "\n",
    "4)\tHeatmap - Draw a heatmap representation of the expression values of the genes in D (from 3d), across the entire cohort (all samples). Order the genes and the samples to produce the maximal visual effect.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
